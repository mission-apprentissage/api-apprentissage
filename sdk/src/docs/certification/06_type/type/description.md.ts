// This file is generated by a script and should not be modified directly.
const markdown: string =
  "Le type de certification renseigne sur\n\n- la nature du diplôme\n- le service responsable de la définition du diplôme\n- le(s) nom(s) et numéro(s) SIRET du ou des certificateur(s) renseigné(s) dans le RNCP\n- le type d’enregistrement au RNCP (de droit ou sur demande)\n- la voie d’accès à la formation (apprentissage, expérience, candidature individuelle, contrat de professionnalisation, formation continue et formation statut élève).\n";
export default markdown;
