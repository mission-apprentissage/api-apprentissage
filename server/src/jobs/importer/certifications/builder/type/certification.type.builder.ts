import { internal } from "@hapi/boom";
import { ICertification } from "api-alternance-sdk";

import { ISourceAggregatedData } from "@/jobs/importer/certifications/builder/certification.builder.js";

export function buildCertificationType(data: ISourceAggregatedData): ICertification["type"] {
  const cfdData = data.bcn?.data ?? null;
  const rncpData = data.france_competence?.data ?? null;

  return {
    certificateurs_rncp:
      rncpData?.certificateurs.map((certificateur) => ({
        siret: certificateur.Siret_Certificateur,
        nom: certificateur.Nom_Certificateur,
      })) ?? null,
    enregistrement_rncp: rncpData?.standard?.Type_Enregistrement ?? null,
    gestionnaire_diplome: cfdData?.GESTIONNAIRE_FORMATION_DIPLOME ?? null,
    nature: {
      cfd:
        cfdData === null
          ? null
          : {
              code: cfdData.NATURE_FORMATION_DIPLOME ?? null,
              libelle: cfdData.N_NATURE_FORMATION_DIPLOME_LIBELLE_100 ?? null,
            },
    },
    voie_acces: {
      rncp:
        rncpData === null
          ? null
          : rncpData.voies_d_acces.reduce(
              (acc, voie) => {
                if (voie.Si_Jury === "En contrat d’apprentissage") {
                  acc.apprentissage = true;
                } else if (voie.Si_Jury === "Par expérience") {
                  acc.experience = true;
                } else if (voie.Si_Jury === "Par candidature individuelle") {
                  acc.candidature_individuelle = true;
                } else if (voie.Si_Jury === "En contrat de professionnalisation") {
                  acc.contrat_professionnalisation = true;
                } else if (voie.Si_Jury === "Après un parcours de formation continue") {
                  acc.formation_continue = true;
                } else if (voie.Si_Jury === "Après un parcours de formation sous statut d’élève ou d’étudiant") {
                  acc.formation_statut_eleve = true;
                } else {
                  throw internal("import.certifications: unexpected voie d'acces value", { voie });
                }

                return acc;
              },
              {
                apprentissage: false as boolean,
                experience: false as boolean,
                candidature_individuelle: false as boolean,
                contrat_professionnalisation: false as boolean,
                formation_continue: false as boolean,
                formation_statut_eleve: false as boolean,
              }
            ),
    },
  };
}
