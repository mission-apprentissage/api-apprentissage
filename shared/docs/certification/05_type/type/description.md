Le type de certification renseigne sur

- la nature du diplôme
- le service responsable de la définition du diplôme
- le(s) nom(s) et numéro(s) SIRET du ou des certificateur(s) renseigné(s) dans le RNCP
- le type d’enregistrement au RNCP (de droit ou sur demande)
- la voie d’accès à la formation (apprentissage, expérience, candidature individuelle, contrat de professionnalisation, formation continue et formation statut élève).
