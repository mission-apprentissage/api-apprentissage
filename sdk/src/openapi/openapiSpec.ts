import { certificationModelOpenapi } from "../models/certification/certification.model.openapi.js";
import { communeModelOpenapi } from "../models/geographie/commune.model.openapi.js";
import { departementModelOpenapi } from "../models/geographie/departement.model.openapi.js";
import { missionLocaleModelOpenapi } from "../models/geographie/mission-locale.model.openapi.js";
import {
  applicationWriteModelOpenapi,
  offerReadModelOpenapi,
  offerWriteModelOpenapi,
  recruiterModelOpenapi,
} from "../models/job/job.model.openapi.js";
import { certificationsRoutesOpenapi } from "../routes/certification/certification.routes.openapi.js";
import { geographieRoutesOpenapi } from "../routes/geographie/geographie.routes.openapi.js";
import { jobRoutesOpenapi } from "../routes/jobs/job.routes.openapi.js";
import { organismeRoutesOpenapi } from "../routes/organisme/organisme.routes.openapi.js";
import { tagsOpenapi } from "./tags.openapi.js";
import type { OpenapiSpec } from "./types.js";

export const openapiSpec: OpenapiSpec = {
  models: {
    [certificationModelOpenapi.name]: certificationModelOpenapi,
    [communeModelOpenapi.name]: communeModelOpenapi,
    [departementModelOpenapi.name]: departementModelOpenapi,
    [missionLocaleModelOpenapi.name]: missionLocaleModelOpenapi,
    [recruiterModelOpenapi.name]: recruiterModelOpenapi,
    [offerReadModelOpenapi.name]: offerReadModelOpenapi,
    [offerWriteModelOpenapi.name]: offerWriteModelOpenapi,
    [applicationWriteModelOpenapi.name]: applicationWriteModelOpenapi,
  },
  routes: {
    ...jobRoutesOpenapi,
    ...certificationsRoutesOpenapi,
    ...geographieRoutesOpenapi,
    ...organismeRoutesOpenapi,
  },
  tags: tagsOpenapi,
};
