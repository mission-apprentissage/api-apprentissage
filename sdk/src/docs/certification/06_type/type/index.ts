import { DocField } from "../../../types.js";
import description from "./description.md.js";
import information from "./information.md.js";

export const typeField: DocField = {
  name: "type",
  description,
  information,
  sample: null,
  tags: [
    ".nature.cfd.code",
    ".nature.cfd.libelle",
    ".gestionnaire_diplome",
    ".certificateurs_rncp[].siret",
    ".certificateurs_rncp[].nom",
    ".enregistrement_rncp",
    ".voie_acces.rncp",
  ],
};
