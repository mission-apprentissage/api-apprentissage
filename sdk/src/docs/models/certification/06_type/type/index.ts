import type { DocBusinessField } from "../../../../types.js";
import description from "./description.md.js";

export default <DocBusinessField>{
  type: "business",
  name: "type",
  description,
  information: null,
  sample: null,
  notes: null,
  tip: null,
  tags: [
    ".nature.cfd.code",
    ".nature.cfd.libelle",
    ".gestionnaire_diplome",
    ".certificateurs_rncp[].siret",
    ".certificateurs_rncp[].nom",
    ".enregistrement_rncp",
    ".voie_acces.rncp",
  ],
};
