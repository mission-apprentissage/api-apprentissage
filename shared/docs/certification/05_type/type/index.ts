import { DocField } from "../../../types";
import description from "./description.md?raw" assert { type: "text" };
import information from "./information.md?raw" assert { type: "text" };

export const typeField: DocField = {
  name: "type",
  description,
  information,
  sample: null,
  tags: [
    ".nature.cfd.code",
    ".nature.cfd.libelle",
    ".gestionnaire_diplome",
    ".certificateurs_rncp[].siret",
    ".certificateurs_rncp[].nom",
    ".enregistrement_rncp",
    ".voie_acces.rncp",
  ],
};
