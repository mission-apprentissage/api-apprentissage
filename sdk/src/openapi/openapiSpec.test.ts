import type { ReferenceObject, SchemaObject } from "openapi3-ts/oas31";
import { describe, expect, it } from "vitest";

import type { DocModel, DocRoute, DocTechnicalField } from "../docs/types.js";
import { openapiSpec } from "./openapiSpec.js";
import type { OpenapiRoute } from "./types.js";

function getDocTechnicalFieldStructure(doc: DocTechnicalField, prefix: string = ""): string[] {
  const structure: string[] = [prefix];

  if (!doc._) {
    return structure;
  }

  for (const key in doc._) {
    if (doc._[key]) {
      structure.push(...getDocTechnicalFieldStructure(doc._[key], `${prefix}.${key}`));
    }
  }

  return structure;
}

function getDocModelStructure(doc: DocModel, prefix: string = ""): string[] {
  const structure: string[] = [prefix];

  if (!doc._) {
    return structure;
  }

  for (const key in doc._) {
    if (doc._[key]) {
      structure.push(...getDocTechnicalFieldStructure(doc._[key], `${prefix}.${key}`));
    }
  }

  return structure.toSorted();
}

function getSchemaObjectStructure(schema: SchemaObject | ReferenceObject | undefined, prefix: string = ""): string[] {
  const structure: string[] = [prefix];

  if (!schema || "$ref" in schema) {
    return structure;
  }

  if (schema.properties) {
    for (const key in schema.properties) {
      if (schema.properties[key]) {
        structure.push(...getSchemaObjectStructure(schema.properties[key], `${prefix}.${key}`));
      }
    }
  }

  if (schema.items) {
    structure.push(...getSchemaObjectStructure(schema.items, `${prefix}.[]`));
  }

  if (schema.prefixItems) {
    for (const key in schema.prefixItems) {
      if (schema.prefixItems[key]) {
        structure.push(...getSchemaObjectStructure(schema.prefixItems[key], `${prefix}.${key}`));
      }
    }
  }

  return structure.toSorted();
}

function getDocRouteStructure(route: DocRoute | null, prefix: string = ""): string[] {
  const structure: string[] = [prefix];

  if (route === null) {
    return structure;
  }

  if (route.parameters) {
    for (const [key, parameter] of Object.entries(route.parameters)) {
      structure.push(...getDocTechnicalFieldStructure(parameter, `${prefix}.parameters.${key}`));
    }
  }

  if (route.body) {
    if (route.body.content) {
      structure.push(...getDocTechnicalFieldStructure(route.body.content, `${prefix}.body`));
    } else {
      structure.push(`${prefix}.body`);
    }
  }

  if (route.response) {
    if (route.response.content) {
      structure.push(...getDocTechnicalFieldStructure(route.response.content, `${prefix}.response`));
    } else {
      structure.push(`${prefix}.response`);
    }
  }

  return structure.toSorted();
}

function getContentObjectStructure(content: SchemaObject, prefix: string = ""): string[] {
  const structure: string[] = [];

  for (const [_key, mediaType] of Object.entries(content)) {
    structure.push(...getSchemaObjectStructure(mediaType.schema, prefix));
  }

  return structure;
}

function getOperationObjectStructure(schema: OpenapiRoute["schema"], prefix: string = ""): string[] {
  const structure: string[] = [prefix];

  if (schema.parameters) {
    for (const parameter of schema.parameters) {
      if (!("$ref" in parameter)) {
        structure.push(`${prefix}.parameters.${parameter.name}`);
      }
    }
  }

  if (schema.requestBody) {
    if (!("$ref" in schema.requestBody)) {
      structure.push(...getContentObjectStructure(schema.requestBody.content, `${prefix}.body`));
    }
  }

  if (schema.responses) {
    for (const [key, response] of Object.entries(schema.responses)) {
      if (key === "default" || key.startsWith("2")) {
        if (response.content) {
          structure.push(...getContentObjectStructure(response.content, `${prefix}.response`));
        } else {
          structure.push(`${prefix}.response`);
        }
      }
    }
  }

  return structure.toSorted();
}

describe("openapiSpec#models", () => {
  it.each(Object.entries(openapiSpec.models))("should have doc in sync with schema %s", (_modelName, model) => {
    if (model.doc === null) {
      return;
    }
    expect(getDocModelStructure(model.doc)).toEqual(getSchemaObjectStructure(model.schema));
  });
});

describe("openapiSpec#routes", () => {
  describe.each(Object.entries(openapiSpec.routes))("route %s", (_routeName, route) => {
    it.each(Object.entries(route))("should have doc in sync with schema %s", (_key, operation) => {
      if (operation.doc === null) {
        return;
      }
      expect(getDocRouteStructure(operation.doc)).toEqual(getOperationObjectStructure(operation.schema));
    });
  });
});
