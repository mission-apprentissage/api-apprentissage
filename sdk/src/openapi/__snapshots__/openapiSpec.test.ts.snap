// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`openapiSpec#models > should generate schema in sync with zod definition 1`] = `
{
  "Commune": {
    "diff": {
      "properties.localisation.properties.bbox.properties.coordinates.items.items.maxItems": {
        "result": 2,
        "type": "added",
      },
      "properties.localisation.properties.bbox.properties.coordinates.items.items.minItems": {
        "result": 2,
        "type": "added",
      },
      "properties.localisation.properties.centre.properties.coordinates.maxItems": {
        "result": 2,
        "type": "added",
      },
      "properties.localisation.properties.centre.properties.coordinates.minItems": {
        "result": 2,
        "type": "added",
      },
      "properties.mission_locale.$ref": {
        "source": "#/components/schemas/MissionLocale",
        "type": "removed",
      },
      "properties.mission_locale.anyOf": {
        "result": [
          {
            "type": "null",
          },
          {
            "$ref": "#/components/schemas/MissionLocale",
          },
        ],
        "type": "added",
      },
    },
    "result": "api",
    "source": "zod",
  },
  "MissionLocale": {
    "diff": {
      "properties.localisation.properties.geopoint.properties.coordinates.maxItems": {
        "result": 2,
        "type": "added",
      },
      "properties.localisation.properties.geopoint.properties.coordinates.minItems": {
        "result": 2,
        "type": "added",
      },
      "type": {
        "result": "object",
        "source": [
          "object",
          "null",
        ],
        "type": "changed",
      },
    },
    "result": "api",
    "source": "zod",
  },
}
`;

exports[`openapiSpec#routes > should generate schema in sync with zod definition 1`] = `
{
  "get:/certification/v1": {
    "diff": {
      "parameters.query:identifiant.cfd.allowEmptyValue": {
        "result": true,
        "type": "added",
      },
      "parameters.query:identifiant.rncp.allowEmptyValue": {
        "result": true,
        "type": "added",
      },
    },
    "result": "api",
    "source": "zod",
  },
  "get:/geographie/v1/mission-locale": {
    "diff": {
      "responses.200.content.application/json.schema.items.$ref": {
        "result": "#/components/schemas/MissionLocale",
        "type": "added",
      },
      "responses.200.content.application/json.schema.items.allOf": {
        "source": [
          {
            "$ref": "#/components/schemas/MissionLocale",
          },
          {
            "type": "object",
          },
        ],
        "type": "removed",
      },
    },
    "result": "api",
    "source": "zod",
  },
}
`;
